### PR TITLE
[charts] Use new text component to avoid tick label overflow on x-axis

### PR DESCRIPTION
**Pull request by alexfauquette**
Part of #10140 and #7801

This PR introduces multiple concepts:

1. the tick selection: You can now select which tick to display. For that, you can provide a function similar to a filtering method

2. the tick label selection: You can display labels on a subset of ticks. The same method, you provide a filtering method. But it is applied on the result of the tick filtering. In other words you can not get a label without a tick.

3. automatic label filtering: By default, we keep all ticks, and we display the first label. Then we display the next label that does not overflow, and so on.

4. The ability to rotate the text

5. Pass all the customization in the `style` object.
  - The advantage is that it feels more natural. You can pass the `textBaseline` as a standard CSS property.
  - The drawback is it mixes some CSS properties that will be passed to the element, and some custom ones such as `textBaseline` or `rotate` that are used for the custom `<ChartsText />` computation and then restricted to a subset of options.

## Tech aspect:

To know if a label overflows (especially if they are rotated), we compute its bounding box, and we get interested by how much do we need to translate them to avoid the overflow (the red line)

For small angles, I approximate them to 0° or 90°.

Hypothesis are: all text boxes are the same height, and angles are between -90° and 90°. Without those hypotheses, computing this minimal gap would require getting the dimension of the next rectangle.

image

Here is a demo where each rectangle is aligned according to the minimal translation to avoid overflow. You can see that when width are different it fails for angles larger than 90° But not sure it is worth fixing it since I assume nobody displays labels rotated by 180°

https://github.com/mui/mui-x/assets/45398769/6fab43fb-4885-49c2-b0d4-6fa3978edf30

### CODE REVIEW

File: docs/data/charts/stacking/StackOrderDemo.tsx [hunk 1]
```diff
@@ -105,22 +104,22 @@ export default function StackOrderDemo() {
...
           height={300}
-          xAxis={[xAxis]}
-          yAxis={[{ min: 0, max: 100 }]}
-          series={modifiedSeries}
-          margin={{ bottom: 70 }}
-          sx={{
-            [`.${axisClasses.bottom}`]: {
-              [`.${axisClasses.tickLabel}`]: {
-                transform: 'rotate(45deg)',
+          xAxis={[
```
Comment (alexfauquette) [inline docs/data/charts/stacking/StackOrderDemo.tsx:107]: Basically here all the styling is moved in the axis properties

File: packages/x-charts/src/ChartsLegend/ChartsLegend.tsx [hunk 2]
```diff
@@ -159,20 +159,24 @@ function DefaultChartsLegend(props: LegendRendererProps) {
...
-      fill: (theme.vars || theme).palette.text.primary,
-      lineHeight: 1,
-      ...inLabelStyle,
-    }),
+    () =>
+      ({
+        ...theme.typography.subtitle1,
+        color: 'inherit',
+        dominantBaseline: 'central',
+        textAnchor: 'start',
```
Comment (alexfauquette) [inline packages/x-charts/src/ChartsLegend/ChartsLegend.tsx:167]: Those properties moved into the `style` prop

**Comment by flaviendelangle** [comment]
:star_struck: That's some impressive evolution!

> But not sure it is worth fixing it since I assume nobody displays labels rotated by 180°

Maybe we can just refuse angles we don't handle correctly (either throwing an error or displaying an console warning).

**Comment by alexfauquette** [comment]
Just noticed an issue remains: When the axis automatically removes some tick labels, it leads to a hydration error. Not sure about how to deal with it. Should the computation of the tick label to remove be done after the first render 🤔

**Comment by flaviendelangle** [comment]
I guess it's because on SSR you don't have a notion of component width so we can't guess the ticks to render.

For me either you refuse the usage of the automatic mode on app that use SSR
Or you have an imperfect fallback where the tick labels changes after hydration (but it causes a flicker so it should probably never be a default behavior, just a fallback)

File: packages/x-charts/src/internals/geometry.ts [hunk 3]
```diff
@@ -9,6 +11,18 @@ const ANGLE_APPROX = 5; // Angle (in deg) for which we approximate the rectangle
...
  */
 export function getMinXTranslation(width: number, height: number, angle: number = 0) {
+  if (process.env.NODE_ENV !== 'production') {
+    if (!warnedOnce && angle > 90 && angle < -90) {
+      warnedOnce = true;
+      console.warn(
+        [
+          `MUI X: It seems you applied an angle larger that 90° to an axis text.`,
+          `This could cause some text overlapping.`,
+          `If you encounter a uscase where it's needed, please open an issue.`,
```
Comment (flaviendelangle) [inline packages/x-charts/src/internals/geometry.ts:21]:
```suggestion
          `If you encounter a usecase where it's needed, please open an issue.`,
```

File: docs/data/charts/axis/axis.md [hunk 4]
```diff
@@ -86,6 +88,43 @@ Here the top axis has a `tickMinStep` of half a day, and the bottom axis a `tick
...
+For example `tickInterval: (value, index) => index % 2 === 0` will show only one label every two ticks.
+
+:::warning
+The `value` and `index` in the argument are those of the ticks, not the axis data.
+:::
+
+By default, ticks are filtered such that their labels do not overlap.
+You can override this behavior with `tickInterval: () => true` which show the tick label for each tick.
+
+In this example we still have the top axis as a reference for the default behavior.
```
Comment (LukasTy) [inline docs/data/charts/axis/axis.md:121]:
```suggestion
In this example, the top axis is a reference for the default behavior.
```

File: docs/data/charts/axis/axis.md [hunk 5]
```diff
@@ -107,7 +146,13 @@ Axes rendering can be further customized. Below is an interactive demonstration
 
 {{"demo": "AxisCustomizationNoSnap.js", "hideToolbar": true, "bg": "inline"}}
 
-### Composition
+### Text Customization
```
Comment (LukasTy) [inline docs/data/charts/axis/axis.md:149]:
```suggestion
### Text customization
```

File: packages/x-charts/src/hooks/useTicks.ts [hunk 6]
```diff
@@ -20,6 +20,14 @@ export interface TickParams {
    * Not supported by categorical axis (band, points).
    */
   tickNumber?: number;
+  /**
+   * Defines which ticks are displayed. Its value can be:
+   * - 'auto' In such case the ticks are computed based on axis sacle and other paramteres.
```
Comment (LukasTy) [inline packages/x-charts/src/hooks/useTicks.ts:25]:
```suggestion
   * - 'auto' In such case the ticks are computed based on axis scale and other parameters.
```

File: packages/x-charts/src/hooks/useTicks.ts [hunk 7]
```diff
@@ -40,13 +48,23 @@ export function getTickNumber(
   return Math.min(maxTicks, Math.max(minTicks, defaultizedTickNumber));
 }
 
+export type TickItemType = {
+  /**
+   * This property is undefined only if it's the tick closing the last band
+   * */
```
Comment (LukasTy) [inline packages/x-charts/src/hooks/useTicks.ts:54]:
```suggestion
   */
```

File: packages/x-charts/src/models/axis.ts [hunk 8]
```diff
@@ -61,15 +61,32 @@ export interface ChartsAxisProps extends TickParams {
   /**
    * The font size of the axis ticks text.
    * @default 12
+   * @deprecated You can us `tickLabelStyle.fontSize` instead.
```
Comment (LukasTy) [inline packages/x-charts/src/models/axis.ts:64]:
```suggestion
   * @deprecated Consider using `tickLabelStyle.fontSize` instead.
```

File: packages/x-charts/src/models/axis.ts [hunk 9]
```diff
@@ -61,15 +61,32 @@ export interface ChartsAxisProps extends TickParams {
...
+   */
+  tickLabelStyle?: ChartsTextProps['style'];
+  /**
+   * The style applied to the axis label.
+   */
+  labelStyle?: ChartsTextProps['style'];
+  /**
+   * Defines which ticks get its label displayed. Its value can be:
+   * - 'auto' In such case, labels are displayed if they do not overlap with the previous one.
+   * - a filtering function of the form (value, index) => boolean. Warning: the index is tick index, not data ones.
```
Comment (LukasTy) [inline packages/x-charts/src/models/axis.ts:78]:
```suggestion
   * - a filtering function of the form `(value, index) => boolean`. Warning: the index is tick index, not data ones.
```

File: packages/x-charts/src/models/axis.ts (hunk 8 above)
Comment (LukasTy) [reply packages/x-charts/src/models/axis.ts:64]: This led me to discover a regression in @mui/docs-infra.
Previously, such use cases used to be transformed into being wrapped in a `<code>` block, whereas now it is wrapped in a `span`. 🙈 
Screenshot 2023-10-18 at 17 37 45

File: packages/x-charts/src/ChartsXAxis/ChartsXAxis.tsx [hunk 10]
```diff
@@ -25,12 +27,61 @@ const useUtilityClasses = (ownerState: ChartsXAxisProps & { theme: Theme }) => {
...
+    const { width, offset, labelOffset, height } = item;
+
+    const distance = getMinXTranslation(width, height, style?.angle);
+    const textPosition = offset + labelOffset;
+    const gapRatio = 1.2; // Ratio applied to the minimal distanc eto add some margin
+
+    textStart = textPosition - (gapRatio * distance) / 2;
+    if (labelIndex > 0 && textStart < textEnd) {
+      // Except for the first label, we skip all label that overlap with the last accepted.
+      // Notice that the early return prevent textEnd to be updated.
```
Comment (LukasTy) [inline packages/x-charts/src/ChartsXAxis/ChartsXAxis.tsx:73]:
```suggestion
      // Notice that the early return prevents `textEnd` from being updated.
```

File: packages/x-charts/src/ChartsXAxis/ChartsXAxis.tsx [hunk 11]
```diff
@@ -63,14 +120,8 @@ function ChartsXAxis(inProps: ChartsXAxisProps) {
 
   const tickSize = disableTicks ? 4 : tickSizeProp;
 
-  const xTicks = useTicks({ scale: xScale, tickNumber, valueFormatter });
   const positionSigne = position === 'bottom' ? 1 : -1;
```
Comment (LukasTy) [inline packages/x-charts/src/ChartsXAxis/ChartsXAxis.tsx:123]:
```suggestion
  const positionSign = position === 'bottom' ? 1 : -1;
```

File: packages/x-charts/src/ChartsXAxis/ChartsXAxis.tsx [hunk 12]
```diff
@@ -118,7 +184,7 @@ function ChartsXAxis(inProps: ChartsXAxisProps) {
         />
       )}
 
-      {xTicks.map(({ formattedValue, offset, labelOffset }, index) => {
+      {xTicksWithDimension.map(({ formattedValue, offset, labelOffset, skipLabel }, index) => {
         const xTickLabel = labelOffset ?? 0;
         const yTickLabel = positionSigne * (tickSize + 3);
```
Comment (LukasTy) [inline packages/x-charts/src/ChartsXAxis/ChartsXAxis.tsx:189]:
```suggestion
        const yTickLabel = positionSign * (tickSize + 3);
```

File: docs/data/charts/axis/axis.md [hunk 13]
```diff
@@ -107,7 +146,13 @@ Axes rendering can be further customized. Below is an interactive demonstration
...
+
+To customize the text elements (ticks label and the axis label) use the `tickLabelStyle` and `labelStyle` properties of the axis configuration.
+
+{{"demo": "AxisTextCustomizationNoSnap.js", "hideToolbar": true, "bg": "inline"}}
```
Comment (LukasTy) [inline docs/data/charts/axis/axis.md:153]: WDYT about the fact that the tick labels almost overlap the tick itself with the default angle of 45 degrees and the more you rotate towards 90 degrees, the worse it gets. 🤔 
Should we maybe include some margin? Or the text management itself should account for it? 🤔 

https://github.com/mui/mui-x/assets/4941090/61d33aef-d2b8-43a6-9ae7-4c03826b37b0

**Review by LukasTy — COMMENTED** [review]
Great work! 👍 

Leaving some nitpick comments, functional change suggestions and documentation improvement proposals.

File: docs/data/charts/axis/axis.md [hunk 14]
```diff
@@ -86,6 +88,43 @@ Here the top axis has a `tickMinStep` of half a day, and the bottom axis a `tick
 
 {{"demo": "TickNumber.js"}}
 
+### Fix tick positions
+
+If you want more control on the tick positions, you can use the `tickInterval` property.
+
+This property accept an array of values.
```
[the remark this answers is not in the document]
Comment (LukasTy) [reply docs/data/charts/axis/axis.md:95]: Then maybe in the next sentence mention that only for this scale type the property can ALSO be a callback..? 🤔

File: docs/data/charts/axis/axis.md [hunk 15]
```diff
@@ -86,6 +88,43 @@ Here the top axis has a `tickMinStep` of half a day, and the bottom axis a `tick
...
+The bottom axis use a filtering function to only display a tick at the beginning of a day.
+
+{{"demo": "TickPosition.js"}}
+
+### Filtering ticks label
+
+You can display labels only on a subsets of the ticks with the `tickInterval` property.
+It's a filtering function `(value, index) => boolean`.
+
+For example `tickInterval: (value, index) => index % 2 === 0` will show only one label every two ticks.
```
[the remark this answers is not in the document]
Comment (alexfauquette) [reply docs/data/charts/axis/axis.md:112]:
```suggestion
For example `tickLabelInterval: (value, index) => index % 2 === 0` will show the label only on one tick out of two.
```

[the remark this answers is not in the document]
Comment (alexfauquette) [reply docs/data/charts/axis/axis.md:112]: I agree but "every other something" is an idiomatic expression so either you know it or not. So for a non native speaker (like me 30 minutes ago) this sentence is useless 😉

The [cambride](https://dictionary.cambridge.org/fr/dictionnaire/anglais/every-other) definition is "not each one in a series, but every two". SO maybe this is a correct and non idiomatic version


```suggestion
For example `tickLabelInterval: (value, index) => index % 2 === 0` will show the label every two ticks.
```

[the remark this answers is not in the document]
Comment (LukasTy) [reply docs/data/charts/axis/axis.md:112]: Fair point and the latest suggestion does make sense to me as well. 👍 💯

File: packages/x-charts/src/internals/geometry.ts (hunk 3 above)
Comment (samuelsycamore) [reply packages/x-charts/src/internals/geometry.ts:21]: 😁✌️

```suggestion
          `If you encounter a use case where it's needed, please open an issue.`,
```

File: packages/x-charts/src/models/axis.ts [hunk 16]
```diff
@@ -61,15 +61,32 @@ export interface ChartsAxisProps extends TickParams {
   /**
    * The font size of the axis ticks text.
    * @default 12
+   * @deprecated Consider using  `tickLabelStyle.fontSize` instead.
```
Comment (LukasTy) [inline packages/x-charts/src/models/axis.ts:64]: Nit: Useless empty space.
```suggestion
   * @deprecated Consider using `tickLabelStyle.fontSize` instead.
```

File: packages/x-charts/src/models/axis.ts [hunk 17]
```diff
@@ -61,15 +61,32 @@ export interface ChartsAxisProps extends TickParams {
...
+   */
+  tickLabelInterval?: 'auto' | ((value: any, index: number) => boolean);
   /**
    * The label of the axis.
    */
   label?: string;
   /**
    * The font size of the axis label.
    * @default 14
+   * @deprecated Consider using  `labelStyle.fontSize` instead.
```
Comment (LukasTy) [inline packages/x-charts/src/models/axis.ts:89]: Nit: Useless empty space.
```suggestion
   * @deprecated Consider using `labelStyle.fontSize` instead.
```

File: docs/data/charts/axis/axis.md (hunk 13 above)
Comment (LukasTy) [reply docs/data/charts/axis/axis.md:153]: Cool, the updated demo looks great! 🙌

File: docs/data/charts/axis/AxisTextCustomizationNoSnap.js [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+        [
+          `import { ScatterChart } from '@mui/x-charts/ScatterChart';`,
+          '',
+          `<ScatterChart`,
+          '  {/** ... */}',
+          `  bottomAxis={{`,
+          `    labelStyle: {`,
+          `      fontSize: ${props.labelFontSize},`,
+          `      transform: \`translateY(\${
+            // Hack that should be adde din the lib latter.
```
Comment (LukasTy) [inline docs/data/charts/axis/AxisTextCustomizationNoSnap.js:155]:
```suggestion
            // Hack that should be added in the lib latter.
```

**Review by LukasTy — APPROVED** [review]
Superb improvement! 💙 🚀

File: packages/x-charts/src/internals/geometry.ts (hunk 3 above)
Comment (alexfauquette) [reply packages/x-charts/src/internals/geometry.ts:21]: considered sold 🧑‍⚖️

**Pull request merged**